packages/jsii-rosetta/lib/rosetta-translator.ts

```diff
@@ -0,0 +1,155 @@
+import * as spec from '@jsii/spec';
+
+import { TypeFingerprinter } from './jsii/fingerprinting';
+import { TARGET_LANGUAGES } from './languages';
+import * as logging from './logging';
+import { TypeScriptSnippet, completeSource } from './snippet';
+import { snippetKey } from './tablets/key';
+import { LanguageTablet, TranslatedSnippet } from './tablets/tablets';
+import { translateAll, TranslateAllResult } from './translate_all';
+
+export interface RosettaTranslatorOptions {
+  /**
+   * Assemblies to use for fingerprinting
+   *
+   * The set of assemblies here are used to invalidate the cache. Any types that are
+   * used in snippets are looked up in this set of assemblies. If found, their type
+   * information is fingerprinted and compared to the type information at the time
+   * compilation of the cached sample. If different, this is considered to be a cache
+   * miss.
+   *
+   * You must use the same set of assemblies when generating and reading the cache
+   * file, otherwise the fingerprint is guaranteed to be different and the cache will
+   * be useless (e.g. if you generate the cache WITH assembly information but
+   * read it without, or vice versa).
+   *
+   * @default No assemblies.
+   */
+  readonly assemblies?: spec.Assembly[];
+
+  /**
+   * Whether to include compiler diagnostics in the compilation results.
+   *
+   * @default false
+   */
+  readonly includeCompilerDiagnostics?: boolean;
+}
+
+/**
+ * Entry point for consumers that want to translate code on-the-fly
+ *
+ * If you want to generate and translate code on-the-fly, in ways that cannot
+ * be achieved by the rosetta CLI, use this class.
+ */
+export class RosettaTranslator {
+  public readonly tablet = new LanguageTablet();
+  private readonly fingerprinter: TypeFingerprinter;
+  private readonly cache = new LanguageTablet();
+  private readonly includeCompilerDiagnostics: boolean;
+
+  public constructor(options: RosettaTranslatorOptions = {}) {
+    this.fingerprinter = new TypeFingerprinter(options?.assemblies ?? []);
+    this.includeCompilerDiagnostics = options.includeCompilerDiagnostics ?? false;
+  }
+
+  public async loadCache(fileName: string) {
+    try {
+      await this.cache.load(fileName);
+    } catch (e) {
+      logging.warn(`Error reading cache ${fileName}: ${e.message}`);
+    }
+  }
+
+  /**
+   * For all the given snippets, try to read translations from the cache
+   *
+   * Will remove the cached snippets from the input array.
+   */
+  public readFromCache(snippets: TypeScriptSnippet[], addToTablet = true): TranslatedSnippet[] {
+    if (!this.cache) {
```

Comment:
this.cache is always instantiated, ya? Should this be something like `this.cache.isEmpty` or `!this.cache.loaded` instead, or just be removed?